Реализация примитивов

gravity182 opened the pull request.

gravity182 committed:
Add spinlock

gravity182 committed:
Убрал херню чтоли

gravity182 committed:
Update

gravity182 committed:
Update

gravity182 committed:
Update

gravity182 committed:
Update

misterixis commented on Spinlock.cpp:
```diff
@@ -0,0 +1,43 @@
+    // одновременно
+    // ------------------
+    // Решение:
+    // 1) Запретить прерывания, но у меня нет доступа к такой команде
```
А язык какую команду использует? Что ты будешь делать в своём абстрактном языке?

blinky-z replied:
писал бы ассемблерный код наверное вручную

gravity182 replied:
или использовать команды предоставляемые процессором - CMPXCHG

misterixis replied:
Что-то не понятно, что ты написал. Как будто, первая часть фразы пропала. Что значит "или"?

Если ты скомпилируешь код, задизасемблишь его, там будет CMPXCHG?

gravity182 replied:
Нельзя сказать гарантировано что там будет, т.к. реализация зависит от архитектуры процессора

gravity182 replied:
Кстати, я смотрел реализацию функций в линуксе (не в языке), и там был команда preempt_disable() и preempt_enable() - это то, что использует линукс для того чтобы на время выполнения куска кода не произошло вытеснения, то есть выключаются прерывания

blinky-z replied:
Как язык сделает такие вещи - действительно зависит от архитектуры

gravity182 replied:
Две основные вещи плюсов, как реализуются в них атомарные операции - это cas операции и использование мьютекса. При чем использование какого именно способа, зависит от архитектуры.
 Если есть поддержка cas на аппаратном уровне, то да, язык заиспользует cas операции, т.к. плюсы стараются делать атомарные операции lock free, если нет поддержки cas на аппаратном уровне, то используется мьютекс

gravity182 replied:
Возможно есть ещё какие то способы кроме cas операций и мьютекса, но я не знаю о них

misterixis replied:
А как реализуется мьютекс? Ты пишешь, что он использутеся, если нет поддержки CAS на аппаратном уровне, интересно, как он сам работает?

gravity182 replied:
я не знаю пока как он работает, не реализовывал еще сам. завтра сделаю его и тогда отвечу

gravity182 replied:
Теперь реализовав мьютекс могу сказать:
Мьютекс реализуется также с использованием cas-операций, ведь для того чтобы только 1 поток смог захватить лок, то есть атомарно прочитать и установить значение, также требуется поддержка cas-операций (test-and-set, compare-and-swap) со стороны процессора.

gravity182 replied:
При чем моя фраза 
> Две основные вещи плюсов, как реализуются в них атомарные операции - это cas операции и использование мьютекса
теперь не является правдой, так как мьютекс тоже реализован с помощью cas-операций

gravity182 replied:
значит атомарные операции при неимении cas операций будут использовать что-то другое, не мьютекс

gravity182 replied:
В общем, теперь ещё раз отвечаю.
Я писал какую-то херню, если нет cas-операций, то используется мьютекс, это правда. Однако, это не тот мьютекс что в либе std <mutex>, который прописан в стандарте, а совершенно другой **мьютекс, реализуемый компилятором для того чтобы реализовать атомарные операции языка.** 
 То есть важно провести границу между **языком** и **компилятором**. Есть атомарные операции, прописанные в стандарте C++ - они должны быть так или иначе сделаны компилятором, если компилятор заявлен как поддерживающий стандарт 11 плюсов. Но то как они реализованы - за это отвечает компилятор. Есть поддержка cas операций на аппаратном уровне? Отлично, компилятор генерирует lock free алгоритм. Нету поддержки cas операций? Окей, компилятор делает свой blocking примитив - мьютекс, и тут воля компилятора как он его сделает. Язык только говорит то что операция должна быть атомарной, но не говорит как реализовывать, остальное - дело компилятора.
 Если cas операции на глубоком уровне я увидел и поизучал немного (ну как, более менее понял: Test-and-Set - это команда `XCHG` на Intel x86, а Compare-and-Swap - `lock cmpxchg`), то мьютекс точно нет смысла изучать, ибо он будет зависеть от компилятора, в отличие от аппаратных команд `compare and swap` и `test-and-set`, которые компилятор использует напрямую и код lock free алгоритмов будет плюс минус похожий на различных компиляторах.
 Правда вот очень непонятна ситуация, как компилятор будет делать std::atomic_flag если нет поддержки cas и test-and-set команд на уровне процессора, а этот класс должен быть гарантированно lock free,

misterixis commented on Spinlock.cpp:
```diff
@@ -0,0 +1,43 @@
+    // ------------------
+    // Решение:
+    // 1) Запретить прерывания, но у меня нет доступа к такой команде
+    // 2) Использовать атомарную переменную самого языка, при чем, реализация атомарной переменной должна быть lock-free,
```
Откуда у нас требование к lock-free? Почему wait-free не подойдёт? Что это вообще за штуки такие? Какая из них даёт бОльшие гарантии?

gravity182 replied:
Lock free требуется потому что сам спинлок этого требует, он не должен блокироавться, а крутиться в цикле

gravity182 replied:
Lock-free свойство гарантирует что хотя бы один поток из всех кто вызывает lock free функцию будет добиваться прогресса (т.е. system wide прогресс)
Wait-free гарантирует что все потоки вызывающие эту функцию будут добиваться прогресса в выполнении

gravity182 replied:
когда я вызываю acquire() хотя бы 1 поток точно захватит лок, это гарантированно, acquire() никак не может быть wait-free

misterixis replied:
Ничего не понял. Какая из гарантий более сильная? Не понял, почему понятия wait/lock-free саязаны с спинлоком.

gravity182 replied:
Потому что спинлок - это lock free лок, то есть только 1 поток который заблокирован на спинлоке, должен добиваться прогресса. Поэтому нам нужен atomic_flag, реализация которого полностью lock free, и не менее важно то, что функция test_and_set - атомарна, а это гарантирует то что проверка и установка нового значения будет реализована как неделимая операция, то есть гарантируется, что другой поток не сможет изменять значение флага (лока), и таким образом исключаются race conditions, когда много потоков пытаются одновременно читать значение лока и изменять его. Сделает это в итоге только 1 поток

gravity182 replied:
Wait free спинлок - сделать невозможно. Так как локи как раз так и требуются для того, чтобы исключать race condition, то есть единовременно должен добиваться прогресса только 1 поток, а остальные ждать

gravity182 replied:
Какая из гарантий более сильная - думаю, wait free, так как лучше же многим потокам добиваться прогресса чем одному, вроде по логике так
 Но я ещё почитаю

misterixis replied:
подытожив, сформулируй, что такое lock free and wait free, и про то, что сильнее. И про то, где что из этого используется, и когда надо использовать lock free, а когда wait free?

blinky-z replied:
Так, существует non-blocking программирование
 Оно также может быть lock free, если гарантируется system-wide прогресс, и wait-free, если гарантируется прогресс каждого потока.
 То есть, существует 3 вида: non-blocking, non-blocking lock-free, non-blocking wait-free
  Все wait free алгоритмы - это lock free алгоритмы, но никакой lock free алгоритм не является wait free, потосу что Lock free алгоритмы гарантирует что если много тредов простаивают то хотя бы один будет делать работу, то есть system-wide прогресс, а wait free гарантирует и system-wide прогресс, и то, что никакой поток не будет простаивать бесполезно.
 Wait free алгоритмы нашли свое применение в системах реального времени потому что каждая операция никогда не должна занимать вечность, а всегда делать работу за конечное число шагов (верхний предел) 
 Такие алгоритмы очень сложно писать правильно, и нельзя сказать, что всегда лучше. Но wait free лучше чем lcok free

gravity182 replied:
Так, существует non-blocking программирование
Также существует дальнейшее продолжение в виде lock free, если гарантируется *system-wide* прогресс, и wait-free, если гарантируется прогресс *каждого* потока.
То есть: non-blocking lock-free, non-blocking wait-free

Принцип lock-free - это:
1) Гарантированный _system-wide_ прогресс
2) Если один вымещается планировщиком, это не может помешать выполнению прогресса другими потоками 

Принцип wait-free - это:
1) Гарантированный прогресс *каждого потока*, то есть прогресс одного потока не должен зависеть от другого, ниже будет пример с cas-лупом где в lock free алгоритме 1 поток, делая прогресс, не дает делать прогресс другим
2)  Если один вымещается планировщиком, это не может помешать выполнению прогресса другими потоками 

Все wait free алгоритмы - это lock free алгоритмы, но никакой lock free алгоритм не является wait free.

Wait free алгоритмы нашли свое применение в системах реального времени потому что каждая операция никогда не должна занимать вечность, а всегда делать работу за конечное число шагов (верхний предел) 

Такие алгоритмы очень сложно писать правильно и тем более wait free. Но wait free всегда лучше чем lock free, так как wait free является starvation-free (потоки не простаивают в ожидании бесполезно, а делают какую то работу)

gravity182 replied:
Пример lock free алгоритма: атомарная операция `fetch_add()` реализованная с помощью cas операций 

Пусть несколько потоков вызвали эту функцию. 
Один поток будет крутиться в cas-цикле, и может крутиться даже бесконечно, но если этот поток не смог совершить операцию полностью (то есть операция compare and swap вернула false), то значит что какой-то другой поток смог выполнить эту операцию полностью (действительно, какой-то поток смог прочитать и установить значение атомарно, а остальные вернули ошибку из-за race condition)
Таким образом, какой-то из потоков гарантированно на каждой итерации выполняет полезную работу, это и называется system-wide прогресс
 
Wait free алгоритм же, выполняет свою работу за конечное число шагов, не зависящее от других потоков (в то время как например, если один поток вызвал функцию fetch_add(), количество щагов, за которое она выполнится, будет зависеть от других потоков)

gravity182 replied:
И как противоложный пример спинлока или операции fetch_add реализовнной с помощью cas операций можно привести мьютекс:
 мьютекс - не lock free лок никогда. Пусть 1 поток захватил лок, и по середине работы с критической областью был вытеснен планировщиком, тогда никакой другой поток не сможет захватить мьютекс, пока первый не высвободит его. Понятно, что:
- Тут не происходит даже system-wide прогресса 
- Все потоки пытающиеся захватить мьютекс будут простаивать. Получается, что первый поток неявно заблокировал другие, а это противоречит понятию lock free

misterixis commented on Spinlock.h:
```diff
@@ -0,0 +1,15 @@
+private:
+    std::atomic_flag lock = ATOMIC_FLAG_INIT;
+public:
+    void acquire();
```
Слушай, понятно же, что такие вещи, как канкуренси программы нужно тестить ;)

В джаве ты бы это сделал с помощью jcstress. Тут тоже надо такое.

Надо убедится, что поведение корректное, что нет дедлково, гонок и прочего. Написание такого теста - тоже сложная задача.

gravity182 replied:
у меня тест был но он какой то фиговый
ваще хз как правильно тестить

gravity182 replied:
добавил тест но он фиговый

misterixis replied:
блин, ну если он фиговйый, то зачем ты его добавил. 

Не нужно заниматся pull request driven development, когда ты кидаешь код, и коллеги его подправляют. PR - это код, который ты должен быть готов отправить на реальный боевой сервер.

misterixis commented on CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.12)
+project(ThreadsLearn)
+
+set(CMAKE_CXX_STANDARD 14)
```
почему `14`?

gravity182 replied:
думал что атомики в 14 появились поменял

gravity182 committed:
Update

misterixis commented on spinlock_test.cpp:
```diff
@@ -0,0 +1,34 @@
+const int THREADS_NUM = 2;
+
+void thread_run(int thread_num) {
+    for (int i = 0; i < 100; i++) {
```
мелкие очень цифры для тестирования мнгогопоточности. Там обычно оперерируют порядками "миллион", или больше

misterixis commented on spinlock_test.cpp:
```diff
@@ -0,0 +1,34 @@
+Spinlock lock;
+int cnt = 0;
+
+const int THREADS_NUM = 2;
```
Побольше тредов давай

gravity182 committed:
remove README

gravity182 committed:
Update test

Increase number of threads and operations per thread

gravity182 committed:
Update CMakeLists

Just better formatting

gravity182 committed:
Rename test to concurrent_increment_test

gravity182 committed:
Add Mutex

gravity182 committed:
Update tests

1. Move tests to folder "tests"
2. Add Catch test lib

gravity182 committed:
Add Mutex

gravity182 committed:
Add Mutex tests

misterixis commented on tests/mutex_tests.cpp:
```diff
@@ -0,0 +1,45 @@
+#define CATCH_CONFIG_MAIN
```
Нужны самые примитивные цесты в том числе - если один тред взял лок, другой взять не может

misterixis commented on Mutex.h:
```diff
@@ -0,0 +1,17 @@
+
+class Mutex {
+private:
+    std::atomic_flag lock = ATOMIC_FLAG_INIT;
```
Давай сделаем ещё одну реализацию - на с++ до версии 11

misterixis commented on Mutex.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "Mutex.h"
+
+void Mutex::acquire() {
+    while (lock.test_and_set()) {
```
почему у тебя именно такое условие?

gravity182 replied:
test_and_set() возвращает значение, которое было до начала операции. Если там было 0, то есть лок свободен, то значит сейчас мы смогли захватить его установив 1. Если вернуло 1, то значит лок уже захвачен каким-то другим потоком. 
 Мы знаем, что мы точно смогли захватить значение если до начала операции было 0, потоуму что это атомарная операция и чтение и установка происходят неделимо, как единая операция,то есть гарантируется то, чтодругие потоки ни читали, ни изменяли значение: 0 вернется только одному потоку (лок свободен), остальные получат 1 (лок захвачен кем-то)

gravity182 committed:
Update Mutex

gravity182 committed:
Change setting lock variable from '0' (int) to 'false' (bool)

gravity182 committed:
Add try_lock() func in Mutex

gravity182 committed:
Add mutual exclusion test

gravity182 committed:
Fix mutual exclusion test

gravity182 committed:
Add comments to test

gravity182 committed:
Add try_lock() func in Spinlock

gravity182 committed:
Refactor tests

1. Remove spinlock_tests
2. Make common tests based on mutex tests

gravity182 committed:
Add Cmake file for tests

gravity182 committed:
Change mutex implementation

1. Remove redundant syscalls
2. Now `acquire()`` func based on `try_lock()`` func

gravity182 committed:
Change spinlock implementation

1. Now `acquire()`` func based on `try_lock()`` func

gravity182 committed:
Remove comments from Spinlock

gravity182 committed:
Make using THREADS_NUM constant instead of magical number in Mutual Exclusion test

gravity182 committed:
Add new test

gravity182 committed:
Update tests

gravity182 committed:
Update tests

gravity182 committed:
Update tests

gravity182 committed:
Replace yield() linux func with c++ func from <thread> lib

gravity182 committed:
Update tests

gravity182 committed:
Update tests

gravity182 committed:
Add sleep into Spinlock acquire() func implementation

gravity182 committed:
Return back concurrent increment test

gravity182 committed:
Update

gravity182 committed:
Make threads strict job order

gravity182 committed:
Replace single mutex with mutexes for every cond variables

gravity182 committed:
Update tests

gravity182 committed:
Replace single tests with separate files for mutex and spinlock

gravity182 committed:
Remove sleep from Spinlock acquire() func

gravity182 committed:
Move comments about cond variables into separate file
